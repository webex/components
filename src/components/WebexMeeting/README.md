# Webex Meeting Component

Webex Meeting component displays the complete, default Webex meeting experience.

<p align="center">
  <img src="./WebexMeeting.gif" alt="Default Webex Meeting" />
</p>

## Preview

To see all the different possible states of the Webex Meeting component, you can run our Storybook:

```shell
  npm start
```

## Embed

1. Create a component adapter from which the data will be retrieved (See [adapters](../../adapters)). For instance:

    ```js
    const jsonAdapter = new WebexJSONAdapter(jsonData);
    ```

2. Create a component instance by passing the meeting ID as a string and an optional function that returns an array
of control names for the meeting. The default control names are set to `['mute-audio', 'mute-video', 'join-meeting]`
if the meeting is inactive and `['mute-audio', 'mute-video', 'leave-meeting']` otherwise.
Ensure that the control names match with the adapter implementation of the controls. You then need to enclose it
within [a data provider](../WebexDataProvider/WebexDataProvider.js) that takes
the [component data adapter](../../adapters/WebexJSONAdapter.js) that we created previously

    ```js
    const controls = (isActive) => isActive ? ['join-meeting'] : ['leave-meeting'];

    <WebexDataProvider adapter={jsonAdapter}>
      <WebexMeeting meetingDestination="meetingDestination" controls?={controls}/>
    </WebexDataProvider>
    ```

The component knows how to manage its data. If anything changes in the data source that the adapter manages,
the component will also update on its own.
