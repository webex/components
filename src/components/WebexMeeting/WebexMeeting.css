.meeting {
  display: flex;
  position: relative;
  justify-content: center;
  align-items: center;
  height: 100%;
  width: 100%;
}

.meeting .controls {
  display: flex;
  position: absolute;
  align-self: flex-end;
  bottom: 3rem;
}
