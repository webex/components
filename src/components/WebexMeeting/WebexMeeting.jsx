import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import {Spinner} from '@momentum-ui/react';

import WebexInMeeting from '../WebexInMeeting/WebexInMeeting';
import WebexInterstitialMeeting from '../WebexInterstitialMeeting/WebexInterstitialMeeting';
import WebexMeetingControl from '../WebexMeetingControl/WebexMeetingControl';
import WebexMeetingControls from '../WebexMeetingControl/WebexMeetingControls';
import {WEBEX_COMPONENTS_CLASS_PREFIX} from '../../constants';
import {useMeetingDestination} from '../hooks';

/**
 * Webex Meeting component displays the default Webex meeting experience.
 *
 * @param {string} props.meetingDestination  ID of the virtual meeting location
 * @param {Array} [props.controls]  Array of control names to display
 * @returns {Object} JSX of the component
 */
export default function WebexMeeting({meetingDestination, controls}) {
  const {ID, remoteVideo} = useMeetingDestination(meetingDestination);
  const isActive = remoteVideo !== null;

  const classBaseName = `${WEBEX_COMPONENTS_CLASS_PREFIX}-meeting`;
  const mainClasses = {
    [classBaseName]: true,
    [`${classBaseName}-active`]: ID && isActive,
  };

  const meetingControls = controls(isActive).map(
    (key) => <WebexMeetingControl key={key} type={key} />,
  );

  let meetingDisplay;

  // Undefined meeting ID means that a meeting has never been set
  if (ID === undefined) {
    meetingDisplay = <Spinner />;
  // A null meeting ID means that a meeting is no longer valid (e.g. user left or kicked out)
  } else if (ID === null) {
    meetingDisplay = "You've successfully left the meeting";
  } else {
    meetingDisplay = (
      <>
        {isActive
          ? <WebexInMeeting meetingID={ID} />
          : <WebexInterstitialMeeting meetingID={ID} />}
        <WebexMeetingControls className="meeting-controls-container" meetingID={ID}>
          {meetingControls}
        </WebexMeetingControls>
      </>
    );
  }

  return (
    <div className={classNames(mainClasses)}>
      {meetingDisplay}
    </div>
  );
}

WebexMeeting.propTypes = {
  meetingDestination: PropTypes.string.isRequired,
  controls: PropTypes.func,
};

WebexMeeting.defaultProps = {
  /**
   * A function that returns an array of control names for the meeting.
   * Control name must match with the adapter implementation of the control.
   *
   * @param {boolean} isActive  Whether or not the meeting is active
   */
  // eslint-disable-next-line no-confusing-arrow
  controls: (isActive) =>
    isActive
      ? ['mute-audio', 'mute-video', 'share-screen', 'leave-meeting']
      : ['mute-audio', 'mute-video', 'join-meeting'],
};
