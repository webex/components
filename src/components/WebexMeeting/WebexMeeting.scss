@import '../../styles/variables';

.#{$WEBEX_COMPONENTS_CLASS_PREFIX}-meeting {
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;

  position: relative;
  height: 100%;
  width: 100%;
  min-height: 25rem;
  min-width: 20rem; // Standard small phone width

  .meeting-controls-container {
    position: absolute;
    bottom: 2rem;

    z-index: 100; // Arbitrary large number
  }

  &-active  {
    .meeting-controls {
      .md-button {
        display: none;
      }

      // Display active controls
      .md-button--red {
        display: block;

        background-color: rgba($color: #A12C23, $alpha: 0.4);
        border: solid 1px #D93829;
        color: #D93829;

        // Assumes "leave" is last child
        &:last-child {
          display: none;
        }
      }
    }
  }

  &-active:hover, &-active:active {
    .meeting-controls {
      .md-button {
        display: block;
      }

      .md-button--red {
        background: #FF5C4A;
        border-color: transparent;
        color: #FFFFFF;
      }
    }
  }
}
