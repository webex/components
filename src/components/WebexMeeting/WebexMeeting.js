import React, {Fragment} from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import {Spinner} from '@momentum-ui/react';

import {WEBEX_COMPONENTS_CLASS_PREFIX} from '../../constants';
import {useMeetingDestination} from '../hooks';

import {WebexInMeeting, WebexInterstitialMeeting, WebexMeetingControl, WebexMeetingControls} from '..';

import './WebexMeeting.scss';

/**
 * Webex Meeting component displays the default Webex meeting experience.
 *
 * @param {string} props.meetingDestination  ID of the virtual meeting location
 * @param {Array} [props.controls]  Array of control names to display
 * @returns {Object} JSX of the component
 */
export default function WebexMeeting({meetingDestination, controls}) {
  const {ID, remoteVideo} = useMeetingDestination(meetingDestination);
  const isActive = remoteVideo !== null;

  const classBaseName = `${WEBEX_COMPONENTS_CLASS_PREFIX}-meeting`;
  const mainClasses = {
    [classBaseName]: true,
    [`${classBaseName}-active`]: ID && isActive,
  };

  const meetingControls = controls(isActive).map((key) => <WebexMeetingControl key={key} type={key} />);

  return (
    <div className={classNames(mainClasses)}>
      {ID ? (
        <Fragment>
          {isActive ? <WebexInMeeting meetingID={ID} /> : <WebexInterstitialMeeting meetingID={ID} />}
          <WebexMeetingControls className="meeting-controls-container" meetingID={ID}>
            {meetingControls}
          </WebexMeetingControls>
        </Fragment>
      ) : (
        <Spinner />
      )}
    </div>
  );
}

WebexMeeting.propTypes = {
  meetingDestination: PropTypes.string.isRequired,
  controls: PropTypes.func,
};

WebexMeeting.defaultProps = {
  /**
   * A function that returns an array of control names for the meeting.
   * Control name must match with the adapter implementation of the control.
   *
   * @param {boolean} isActive  Whether or not the meeting is active
   */
  // eslint-disable-next-line no-confusing-arrow
  controls: (isActive) =>
    isActive ? ['mute-audio', 'mute-video', 'leave-meeting'] : ['mute-audio', 'mute-video', 'join-meeting'],
};
