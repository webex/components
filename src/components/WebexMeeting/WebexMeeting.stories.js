import React from 'react';
import WebexMeeting from './WebexMeeting';

export default {
  title: 'Meetings/Webex Meeting',
  component: WebexMeeting,
};

const wrapperStyles = {
  height: '500px',
  width: '650px',
};
const Template = (args) => <div style={wrapperStyles}><WebexMeeting {...args} /></div>;

export const Interstitial = Template.bind({});
Interstitial.args = {
  meetingDestination: 'meeting3',
};

export const Loading = Template.bind({});
Loading.args = {
  meetingDestination: 'meeting9',
};

export const WaitingForOthers = Template.bind({});
WaitingForOthers.args = {
  meetingDestination: 'meeting4',
};

export const InMeeting = Template.bind({});
InMeeting.args = {
  meetingDestination: 'meeting1',
};

export const LeftMeeting = Template.bind({});
LeftMeeting.args = {
  meetingDestination: 'meeting8',
};

export const CustomControls = Template.bind({});
CustomControls.args = {
  meetingDestination: 'meeting3',
  controls: (isActive) => (isActive ? ['leave-meeting'] : ['join-meeting']),
};
