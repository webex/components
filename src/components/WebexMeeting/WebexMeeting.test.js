import React from 'react';

import WebexMeeting from './WebexMeeting';

jest.mock('../hooks/useMeetingDestination');

describe('Webex Meeting component', () => {
  describe('snapshot', () => {
    let controls;

    beforeAll(() => {
      controls = (isActive) => (isActive ? ['leave-meeting'] : ['join-meeting']);
    });

    test('matches snapshot of loading while meeting is created', () => {
      expect(shallow(<WebexMeeting meetingDestination="" />)).toMatchSnapshot();
    });

    test('matches snapshot of a user waiting to join a meeting', () => {
      expect(shallow(<WebexMeeting meetingDestination="localMedia" />)).toMatchSnapshot();
    });

    test('matches snapshot of a user that has joined a meeting', () => {
      expect(shallow(<WebexMeeting meetingDestination="remoteMedia" />)).toMatchSnapshot();
    });

    test('matches snapshot of a user waiting to join a meeting with optional controls', () => {
      expect(shallow(<WebexMeeting meetingDestination="localMedia" controls={controls} />)).toMatchSnapshot();
    });

    test('matches snapshot of a user that has joined a meeting with optional controls', () => {
      expect(shallow(<WebexMeeting meetingDestination="remoteMedia" controls={controls} />)).toMatchSnapshot();
    });
  });
});
