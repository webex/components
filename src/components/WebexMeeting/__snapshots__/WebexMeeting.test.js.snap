// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Webex Meeting component snapshot matches snapshot of a user that has joined a meeting 1`] = `
<div
  className="meeting meeting-active"
>
  <WebexInMeeting
    meetingID="remoteMedia"
  />
  <WebexMeetingControls
    meetingID="remoteMedia"
  >
    <WebexMeetingControl
      key="mute-audio"
      type="mute-audio"
    />
    <WebexMeetingControl
      key="mute-video"
      type="mute-video"
    />
    <WebexMeetingControl
      key="leave-meeting"
      type="leave-meeting"
    />
  </WebexMeetingControls>
</div>
`;

exports[`Webex Meeting component snapshot matches snapshot of a user that has joined a meeting with optional controls 1`] = `
<div
  className="meeting meeting-active"
>
  <WebexInMeeting
    meetingID="remoteMedia"
  />
  <WebexMeetingControls
    meetingID="remoteMedia"
  >
    <WebexMeetingControl
      key="leave-meeting"
      type="leave-meeting"
    />
  </WebexMeetingControls>
</div>
`;

exports[`Webex Meeting component snapshot matches snapshot of a user waiting to join a meeting 1`] = `
<div
  className="meeting"
>
  <WebexInterstitialMeeting
    meetingID="localMedia"
  />
  <WebexMeetingControls
    meetingID="localMedia"
  >
    <WebexMeetingControl
      key="mute-audio"
      type="mute-audio"
    />
    <WebexMeetingControl
      key="mute-video"
      type="mute-video"
    />
    <WebexMeetingControl
      key="join-meeting"
      type="join-meeting"
    />
  </WebexMeetingControls>
</div>
`;

exports[`Webex Meeting component snapshot matches snapshot of a user waiting to join a meeting with optional controls 1`] = `
<div
  className="meeting"
>
  <WebexInterstitialMeeting
    meetingID="localMedia"
  />
  <WebexMeetingControls
    meetingID="localMedia"
  >
    <WebexMeetingControl
      key="join-meeting"
      type="join-meeting"
    />
  </WebexMeetingControls>
</div>
`;

exports[`Webex Meeting component snapshot matches snapshot of loading while meeting is created 1`] = `
<div
  className="meeting"
>
  <Spinner
    className=""
    color="black"
    percentage={null}
    showCheck={false}
    showPercentage={false}
    size={36}
  />
</div>
`;
